staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -461,3 +539,148 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
 
 	return nil
 }
+
+// Helper function to create a set of Device instances.
+func createDeviceSet(devices []resourceapi.Device) map[string]struct{} {
+	deviceSet := make(map[string]struct{})
+	for _, device := range devices {
+		deviceSet[device.Name] = struct{}{}
+	}
+	return deviceSet
+}
+
+// getDevicesToAddAndDelete calculates the devices to add and delete from the pool.
+func getDevicesToAddAndDelete(pool Pool, slices []*resourceapi.ResourceSlice) ([]resourceapi.Device, []resourceapi.Device) {
+	existingDeviceSet := createDeviceSet(pool.Devices)
+	currentDeviceSet := make(map[string]struct{})
+	for _, slice := range slices {
+		for _, device := range slice.Spec.Devices {
+			currentDeviceSet[device.Name] = struct{}{}
+		}
+	}
+
+	needToAdd := []resourceapi.Device{}
+	needToDelete := []resourceapi.Device{}
+
+	for _, device := range pool.Devices {
+		if _, exists := currentDeviceSet[device.Name]; !exists {
+			needToAdd = append(needToAdd, device)
+		}
+	}
+
+	for _, slice := range slices {
+		for _, device := range slice.Spec.Devices {
+			if _, exists := existingDeviceSet[device.Name]; !exists {
+				needToDelete = append(needToDelete, device)
+			}
+		}
+	}
+
+	return needToAdd, needToDelete
+}
+
+// removeDevicesFromSlice removes the devices from the slice.
+func removeDevicesFromSlice(slices []*resourceapi.ResourceSlice, devicesToDelete []resourceapi.Device) []*resourceapi.ResourceSlice {
+	newSlices := make([]*resourceapi.ResourceSlice, len(slices))
+	for i, slice := range slices {
+		newDevices := make([]resourceapi.Device, 0, len(slice.Spec.Devices))
+		for _, device := range slice.Spec.Devices {
+			if !containsDevice(devicesToDelete, device) {
+				newDevices = append(newDevices, device)
+			}
+		}
+
+		newSlice := &resourceapi.ResourceSlice{
+			ObjectMeta: slice.ObjectMeta,
+			Spec: resourceapi.ResourceSliceSpec{
+				Driver: slice.Spec.Driver,
+				Pool: resourceapi.ResourcePool{
+					Name:       slice.Spec.Pool.Name,
+					Generation: slice.Spec.Pool.Generation,
+				},
+				NodeName:     slice.Spec.NodeName,
+				NodeSelector: slice.Spec.NodeSelector,
+				AllNodes:     slice.Spec.AllNodes,
+				Devices:      newDevices,
+			},
+		}
+		newSlices[i] = newSlice
+	}
+	return newSlices
+}
+
+// mergeSlices merges the given ResourceSlice instances.
+// It returns the merged ResourceSlice instances and the ResourceSlice instances that need to be deleted.
+func mergeSlices(slices []*resourceapi.ResourceSlice, maxDevicesPerSlice int) ([]*resourceapi.ResourceSlice, []*resourceapi.ResourceSlice) {
+	var needToDeleteSlice []*resourceapi.ResourceSlice
+	for i := 0; i < len(slices); {
+		//if len(slices[i].Spec.Devices) == 0 {
```

Comment:
At that time, I added this logic. When the internal device is empty, it is added to the list of slices to be deleted. https://github.com/kubernetes/kubernetes/pull/127645/files#r1804903267 After the modification, the CI failure scenario will not recur, but I will still carefully check whether there are other scenarios that I have not considered.